primitives/src/pcs/univariate_kzg/mod.rs

```diff
@@ -910,4 +1247,122 @@ mod tests {
+                        degree, rng,
+                    )
+                })
+                .collect();
+            let _comms =
+                <UnivariateKzgPCS<E> as GPUCommit<E, C>>::batch_commit_with_gpu(&ck, &polys)?;
+
+            Ok(())
+        }
+
+        fn test_gpu_e2e_template<E, C>() -> Result<(), PCSError>
+        where
+            C: IcicleCurve + MSM<C>,
+            C::ScalarField: ArkConvertible<ArkEquivalent = E::ScalarField>,
+            C::BaseField:
+                ArkConvertible<ArkEquivalent = <C::ArkSWConfig as CurveConfig>::BaseField>,
+            <C::ArkSWConfig as CurveConfig>::BaseField: PrimeField,
+            E: Pairing<G1Affine = Affine<<C as IcicleCurve>::ArkSWConfig>>,
+            UnivariateKzgPCS<E>: GPUCommit<E, C>,
+        {
+            let rng = &mut test_rng();
+            let supported_degree = 2usize.pow(12);
+            let pp = UnivariateKzgPCS::<E>::gen_srs_for_testing(rng, supported_degree)?;
+
+            // testing on smaller degree for correctness
+            for _ in 0..10 {
+                let degree = usize::rand(rng) % 1025;
+                let (ck, vk) = pp.trim(degree)?;
+                let p =
+                    <DensePolynomial<E::ScalarField> as DenseUVPolynomial<E::ScalarField>>::rand(
+                        degree, rng,
+                    );
+                let comm_gpu = <UnivariateKzgPCS<E> as GPUCommit<E, C>>::commit_with_gpu(&ck, &p)?;
+                let comm_cpu = UnivariateKzgPCS::<E>::commit(&ck, &p)?;
+                assert_eq!(comm_gpu, comm_cpu);
+
+                let point = E::ScalarField::rand(rng);
+                let (proof, value) = UnivariateKzgPCS::<E>::open(&ck, &p, &point)?;
+                assert!(
+                    UnivariateKzgPCS::<E>::verify(&vk, &comm_gpu, &point, &value, &proof)?,
+                    "proof was incorrect for max_degree = {}, polynomial_degree = {}",
+                    degree,
+                    p.degree(),
+                );
+
+                // batch commit
+                let batch_size = rng.gen_range(10..100);
+                let polys: Vec<_> = (0..batch_size).map(|_| <DensePolynomial<E::ScalarField> as DenseUVPolynomial<E::ScalarField>>::rand(
+                        degree, rng,
+                    )).collect();
+                let comms_gpu =
+                    <UnivariateKzgPCS<E> as GPUCommit<E, C>>::batch_commit_with_gpu(&ck, &polys)?;
+                let comms_cpu = UnivariateKzgPCS::<E>::batch_commit(&ck, &polys)?;
+                assert_eq!(comms_gpu, comms_cpu);
+            }
+            Ok(())
+        }
+
+        #[test]
+        fn test_gpu_e2e() {
+            test_gpu_e2e_template::<Bn254, IcicleBn254>().unwrap();
+            // testing on large degree for profiling
+            #[cfg(feature = "kzg-print-trace")]
+            gpu_profiling::<Bn254, IcicleBn254>().unwrap();
```

Comment:
Should we move it to another test fn

Reply:
sounds good, will do

Reply:
moved to new test in [ae6a9ca](https://github.com/EspressoSystems/jellyfish/pull/498/commits/ae6a9cadc4fa54aef719af0cce53e5a8b075bbfb)